packages/backend/scripts/diffc.py

```diff
@@ -0,0 +1,241 @@
+import os
+import argparse
+import pty
+
+RED = "\033[31m"
+GREEN = "\033[32m"
+YELLOW = "\033[33m"
+BOLD = "\033[1m"
+RESET = "\033[0m"
+
+# get terminal column size
+terminal_size = os.get_terminal_size()
+terminal_width = terminal_size.columns
+
+# List of keywords to ignore in directory names
+IGNORE_KEYWORDS = ["Multisig", "AddressManager", "ProxyAdmin", "Gnosis"]
+
+
+def get_project_subpath(base_path: str, folder_name: str, directory: str) -> (str, str) or (None, None):
+    possible_directories = {
+        "main_contracts": "implementation/contracts",
+        "bedrock_contracts": "implementation/optimism/packages/contracts-bedrock/contracts",
+        # Add more paths here...
+    }
+
+    for logical_name, subpath in possible_directories.items():
+        full_path: str = os.path.join(
+            base_path, "discovery", folder_name, "ethereum", ".code", directory, subpath)
+        if os.path.exists(full_path):
+            return subpath, logical_name
+
+    return None, None
+
+
+def list_directories(folder_name):
+    base_path = ".."
+    path = os.path.join(base_path, "discovery",
+                        folder_name, "ethereum", ".code")
+
+    try:
+        entries = os.listdir(path)
+        directories = [
+            entry
+            for entry in entries
+            if os.path.isdir(os.path.join(path, entry))
+            and not any(keyword in entry for keyword in IGNORE_KEYWORDS)
+        ]
+        ignored_directories = [
+            entry for entry in entries if any(keyword in entry for keyword in IGNORE_KEYWORDS)
+        ]
+
+        return set(directories), set(ignored_directories)
+    except FileNotFoundError:
+        print(f"The directory {path} does not exist.")
+        return set(), set()
+    except NotADirectoryError:
+        print(f"{path} is not a directory.")
+        return set(), set()
+
+
+def diff_proxies(folder1, folder2, common_directories):
+    base_path = ".."
+    print(BOLD + "\nComparing proxies..." + RESET)
+    no_changes = True
+
+    for directory in common_directories:
+        path1 = os.path.join(base_path, "discovery", folder1,
+                             "ethereum", ".code", directory, "proxy")
+        path2 = os.path.join(base_path, "discovery", folder2,
+                             "ethereum", ".code", directory, "proxy")
+
+        if not os.path.exists(path1):
+            print(f"'proxy' subdirectory does not exist in {path1}")
+            continue
+        if not os.path.exists(path2):
+            print(f"'proxy' subdirectory does not exist in {path2}")
+            continue
+
+        sol_files1 = [os.path.join(root, f) for root, dirs, files in os.walk(
```

Comment:
This is duplicate code with the function below. A good rule of thumb for pulling things out into functions is 
> If I repeat myself three times, it has very good grounds to be promoted to a function